## Building a Recommendation System for Movie Rating

### Importing Librairies

In [84]:
import pandas as pd
import numpy as np
import math

### Loading the dataset

In [85]:
ratings = pd.read_csv("C:\\DATA612\\Movie_Ratings.csv")

### Inspecting the dataset

In [113]:
print(ratings.head())

     Unnamed: 0  Patrick C  Heather  Bryan  Patrick T  Thomas  aaron  vanessa  \
0         Alien        NaN      NaN    2.0        NaN     5.0    4.0      NaN   
1        Avatar        4.0      5.0    5.0        4.0     2.0    NaN      4.0   
2  Blade Runner        5.0      NaN    NaN        NaN     5.0    4.0      NaN   
3    Braveheart        4.0      NaN    5.0        NaN     4.0    4.0      3.0   
4     Dodgeball        5.0      4.0    3.0        2.0     4.0    NaN      4.0   

   greg  brian  ...  Zak  Matt  Chris.1  Josh  Amy  Valerie  Gary  Stephen  \
0   NaN    4.0  ...  NaN   NaN      4.0   3.0  NaN      NaN   2.0      5.0   
1   3.0    NaN  ...  5.0   NaN      NaN   4.0  3.0      2.0   1.0      4.0   
2   1.0    5.0  ...  NaN   NaN      3.0   NaN  3.0      3.0   1.0      NaN   
3   4.0    4.0  ...  5.0   NaN      4.0   NaN  3.0      4.0   5.0      5.0   
4   5.0    3.0  ...  3.0   NaN      3.0   NaN  4.0      3.0   4.0      3.0   

   Jessica  Jeff  
0      NaN   4.0  
1     

In [87]:
ratings_mat = np.array(ratings.iloc[:,1:])

In [114]:
print(ratings_mat[:10,:12])

[[nan nan  2. nan  5.  4. nan nan  4. nan nan nan]
 [ 4.  5.  5.  4.  2. nan  4.  3. nan  3.  5.  4.]
 [ 5. nan nan nan  5.  4. nan  1.  5.  5. nan nan]
 [ 4. nan  5. nan  4.  4.  3.  4.  4. nan  3.  4.]
 [ 5.  4.  3.  2.  4. nan  4.  5.  3.  4.  5.  5.]
 [ 4.  5.  4.  3.  3. nan  4.  5.  5.  5.  5.  5.]
 [nan  5.  5. nan  4.  4.  4.  5.  5.  5.  3.  4.]
 [nan  5.  4. nan  4. nan  2.  3.  3.  1. nan  3.]
 [ 2. nan  3.  5.  2. nan nan nan  1. nan  2.  2.]
 [ 4.  4.  3.  2. nan  5.  2.  3.  5.  2.  4.  4.]]


### Splitting the data into train and test set

In [89]:
rating = ratings_mat.copy()

In [90]:
rating_test = ratings_mat[5:16,5:16].copy()
rating[5:16,5:16]= 'nan' 
rating_train = rating

In [115]:
print(rating_train[5:16,5:16])

[[nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]
 [nan nan nan nan nan nan nan nan nan nan nan]]


In [116]:
print(rating_test)

[[nan  4.  5.  5.  5.  5.  5.  5.  5.  3.  5.]
 [ 4.  4.  5.  5.  5.  3.  4.  3. nan  3.  5.]
 [nan  2.  3.  3.  1. nan  3.  2.  3.  2. nan]
 [nan nan nan  1. nan  2.  2.  2.  1.  1.  1.]
 [ 5.  2.  3.  5.  2.  4.  4.  3.  4.  3.  4.]
 [ 1.  4.  4.  4.  4.  2.  3.  3.  2.  1.  1.]
 [nan  3.  5.  4.  4. nan  5.  4. nan  5.  5.]
 [nan  1. nan  1.  1. nan nan  2. nan  3. nan]
 [ 3. nan  4.  5.  4. nan  5.  4. nan  5.  5.]
 [ 4.  4.  4. nan  4. nan  4. nan nan  3.  3.]
 [ 5. nan  5.  5.  5. nan nan  4. nan  4. nan]]


### Mean of train set

In [93]:
training_mean = np.nanmean(rating_train)
print(training_mean)

3.4971428571428573


### Creating a rating matrix of mean for training and test

In [94]:
rating_mean_mat_test = np.ones(rating_test.shape)*training_mean

In [95]:
rating_mean_mat_train = np.ones(rating_train.shape)*training_mean

### Calculating the RMSE for raw average for both training data and test data.

In [96]:
rmse1_test = math.sqrt(np.nanmean((rating_test - rating_mean_mat_test)**2))

In [97]:
print(rmse1_test)

1.3373015778814


In [98]:
rmse1_train = math.sqrt(np.nanmean((rating_train - rating_mean_mat_train)**2))

In [99]:
print(rmse1_train)

1.2867424460430998


### Calculating the bias for each item.

In [100]:
bias_movies = np.nanmean(rating_train, axis=0) - training_mean

In [101]:
print(bias_movies)

[ 0.11396825  0.43619048 -0.13350649 -0.36380952  0.18467532  0.25285714
 -0.06857143  0.39174603  0.42593407  0.50285714  1.07428571  0.72507937
 -0.24714286  0.21714286 -0.72791209  0.32103896 -0.83047619 -0.31532468
  0.68467532 -0.43831933 -0.10583851 -0.44952381 -0.16380952 -0.24714286
  0.13922078]


### Calculate the bias for each user.

In [102]:
bias_users = np.nanmean(rating_train, axis=1) - training_mean

In [103]:
print(bias_users)

[ 0.00285714  0.05285714  0.25285714  0.39759398  0.12190476  0.36
  0.39174603  0.32103896 -1.60825397 -0.13350649 -0.72791209  0.20285714
 -1.49714286  0.16952381  0.10285714  0.83619048 -1.36380952  0.00285714
  1.1115528   0.92390977 -2.37214286  0.58981366  0.78285714 -0.96380952
 -1.76380952]


### Calculate the RMSE for the baseline predictors for both your training data and your test data.

## Baseline Predictor = Raw avg + Bias users + Bias Movies

In [104]:
baseline = np.zeros((25, 25))
for i in range(len(bias_users)):
    for j in range(len(bias_movies)):
        baseline[i,j] = training_mean + bias_users[i] + bias_movies[j]
        if baseline[i,j] < 1 :
            baseline[i,j] = 1
        if baseline[i,j] > 5:
            baseline[i,j] = 5

In [112]:
print(baseline)

[[3.61396825 3.93619048 3.36649351 3.13619048 3.68467532 3.75285714
  3.43142857 3.89174603 3.92593407 4.00285714 4.57428571 4.22507937
  3.25285714 3.71714286 2.77208791 3.82103896 2.66952381 3.18467532
  4.18467532 3.06168067 3.39416149 3.05047619 3.33619048 3.25285714
  3.63922078]
 [3.66396825 3.98619048 3.41649351 3.18619048 3.73467532 3.80285714
  3.48142857 3.94174603 3.97593407 4.05285714 4.62428571 4.27507937
  3.30285714 3.76714286 2.82208791 3.87103896 2.71952381 3.23467532
  4.23467532 3.11168067 3.44416149 3.10047619 3.38619048 3.30285714
  3.68922078]
 [3.86396825 4.18619048 3.61649351 3.38619048 3.93467532 4.00285714
  3.68142857 4.14174603 4.17593407 4.25285714 4.82428571 4.47507937
  3.50285714 3.96714286 3.02208791 4.07103896 2.91952381 3.43467532
  4.43467532 3.31168067 3.64416149 3.30047619 3.58619048 3.50285714
  3.88922078]
 [4.0087051  4.33092732 3.76123035 3.53092732 4.07941217 4.14759398
  3.82616541 4.28648287 4.32067091 4.39759398 4.96902256 4.61981621
  3.64

In [106]:
rmse2_test = math.sqrt(np.nanmean((rating_test - baseline[5:16, 5:16])**2))

In [107]:
print(rmse2_test)

1.0177378348800252


In [108]:
rmse2_train = math.sqrt(np.nanmean((rating_train - baseline)**2))

In [109]:
print(rmse2_train)

0.9139315000290194


### mesure of improvement

In [110]:
print((rmse2_test - rmse1_test)/rmse1_test)

-0.2389616136605767


### There is an improvement of 24% on the test set from the raw average model to the Baseline Model

In [111]:
print((rmse2_train - rmse1_train)/rmse1_train)

-0.28973237586163614


### There is an improvement of 29% on the train set from raw average model to Baseline Model